### PR TITLE
unixODBCDrivers.redshift: init at 1.4.49.1000

### PR DESCRIPTION
**Pull request by sir4ur0n**
- [x] Prerequisite: https://github.com/NixOS/nixpkgs/pull/168177 is merged

Add Amazon Redshift ODBC driver.

As Amazon does not provide sources, we extract the Debian package and manually patchelf.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by thufschmitt — APPROVED** [review]

File: pkgs/development/libraries/unixODBCDrivers/default.nix [hunk 1]
```diff
@@ -174,4 +174,45 @@
       maintainers = with maintainers; [ spencerjanssen ];
     };
   };
+
+  redshift = stdenv.mkDerivation rec {
+    pname = "redshift-odbc";
+    version = "1.4.45.1000-1";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/unixODBCDrivers/default.nix:180]:
```suggestion
    version = "1.4.45.1000";
```

File: pkgs/development/libraries/unixODBCDrivers/default.nix [hunk 2]
```diff
@@ -174,4 +174,45 @@
...
+
+    src = fetchurl {
+      url = "https://s3.amazonaws.com/redshift-downloads/drivers/odbc/1.4.45.1000/AmazonRedshiftODBC-64-bit-${version}.x86_64.deb";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/unixODBCDrivers/default.nix:183]:
```suggestion
      url = "https://s3.amazonaws.com/redshift-downloads/drivers/odbc/${version}/AmazonRedshiftODBC-64-bit-${version}-1.x86_64.deb";
```

File: pkgs/development/libraries/unixODBCDrivers/default.nix [hunk 3]
```diff
@@ -174,4 +174,45 @@
...
+      cd src
+    '';
+
+    # `unixODBC` is loaded with `dlopen`, so `autoPatchElfHook` cannot see it, and `patchELF` phase would strip the manual patchelf. Thus:
+    # - Manually patchelf with `unixODCB` libraries
+    # - Disable automatic `patchELF` phase
+    installPhase = ''
+      mkdir -p $out/lib
+      cp opt/amazon/redshiftodbc/lib/64/* $out/lib
+      ${patchelf}/bin/patchelf --set-rpath ${unixODBC}/lib/ $out/lib/libamazonredshiftodbc64.so
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/unixODBCDrivers/default.nix:198]:
```suggestion
      patchelf --set-rpath ${unixODBC}/lib/ $out/lib/libamazonredshiftodbc64.so
```
part of stdenv

File: pkgs/development/libraries/unixODBCDrivers/default.nix [hunk 4]
```diff
@@ -174,4 +174,45 @@
...
+    pname = "redshift-odbc";
+    version = "1.4.45.1000-1";
+
+    src = fetchurl {
+      url = "https://s3.amazonaws.com/redshift-downloads/drivers/odbc/1.4.45.1000/AmazonRedshiftODBC-64-bit-${version}.x86_64.deb";
+      sha256 = "sha256-WCntyp+Him8huzTlIE/tPwnzBj6zyw7A3S8mz1gqeTU=";
+    };
+
+    unpackPhase = ''
+      ${dpkg}/bin/dpkg -x $src src
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/unixODBCDrivers/default.nix:188]:
```suggestion
      dpkg -x $src src
```
Please add it to nativeBuildInputs

File: pkgs/development/libraries/unixODBCDrivers/default.nix [hunk 5]
```diff
@@ -174,4 +174,45 @@
...
+
+    dontPatchELF = true;
+
+    buildInputs = [ unixODBC ];
+
+    passthru =
+      {
+        fancyName = "Amazon Redshift (x64)";
+        driver = "lib/libamazonredshiftodbc64.so";
+      };
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/unixODBCDrivers/default.nix:209]:
```suggestion
    passthru = {
      fancyName = "Amazon Redshift (x64)";
      driver = "lib/libamazonredshiftodbc64.so";
    };
```

File: pkgs/development/libraries/unixODBCDrivers/default.nix [hunk 6]
```diff
@@ -174,4 +174,45 @@
...
+
+    meta = with lib; {
+      description = "Amazon Redshift ODBC driver";
+      homepage = "https://docs.aws.amazon.com/redshift/latest/mgmt/configure-odbc-connection.html";
+      license = licenses.unfree;
+      platforms = platforms.linux;
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/unixODBCDrivers/default.nix:215]: Missing maintainer

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by sir4ur0n** [comment]
Thank you @SuperSandro2000 for the thorough and qualitative review!

I'm a bit busy these days but I added this to my todolist, I'll fix all threads in the coming days hopefully!

**Comment by sir4ur0n** [comment]
@SuperSandro2000 I think I addressed all your threads.

On top of that:
* I updated the version to 1.4.49.1000
* I updated the SHA accordingly
* I reworded the commit

Please review again :pray:

File: pkgs/development/libraries/unixODBCDrivers/default.nix [hunk 7]
```diff
@@ -174,4 +174,47 @@
...
+  redshift = stdenv.mkDerivation rec {
+    pname = "redshift-odbc";
+    version = "1.4.49.1000";
+
+    src = fetchurl {
+      url = "https://s3.amazonaws.com/redshift-downloads/drivers/odbc/${version}/AmazonRedshiftODBC-64-bit-${version}-1.x86_64.deb";
+      sha256 = "sha256-r5HvsZjB7+x+ClxtWoONkE1/NAbz90NbHfzxC6tf7jA=";
+    };
+
+    nativeBuildInputs = [ dpkg patchelf ];
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/unixODBCDrivers/default.nix:187]:
```suggestion
    nativeBuildInputs = [ dpkg ];
```
part of stdenv

Comment (sir4ur0n) [reply pkgs/development/libraries/unixODBCDrivers/default.nix]: Done in the latest version

**Pull request merged**